static/src/javascripts/projects/common/modules/ui/bannerPicker.js

```diff
@@ -0,0 +1,81 @@
+// @flow
+import ophan from 'ophan/ng';
+
+export type Banner = {
+    id: string,
+    canShow: () => Promise<boolean>,
+    show: () => void,
+};
+
+const init = (banners: Array<Banner>): Promise<void> => {
+    const results: Array<'pending' | boolean> = new Array(banners.length).fill(
+        'pending',
+        0
+    );
+
+    const getSuccessfulBannerIndex = (): number => {
+        const firstCheckPassedIndex = results.findIndex(item => item === true);
+
+        // if no check has passed firstCheckPassedIndex equals -1
+        // if first check has passed firstCheckPassedIndex equals 0
+        if (firstCheckPassedIndex <= 0) {
+            return firstCheckPassedIndex;
+        }
+
+        // if firstCheckPassedIndex greater than 0 then get higher priority checks from array that are pending
+        const pendingHigherPriorityCheckIndex = results
+            .slice(0, firstCheckPassedIndex)
+            .findIndex(item => item === 'pending');
+
+        // if there are no higher priority checks pending return firstCheckPassedIndex
+        if (pendingHigherPriorityCheckIndex === -1) {
+            return firstCheckPassedIndex;
+        }
+
+        return -1;
+    };
+
+    return new Promise(resolve => {
+        const TIME_LIMIT = 2000;
+
+        banners.forEach((banner, index) => {
+            const pushToResults = (result: boolean): void => {
+                results[index] = result;
+
+                const successfulBannerIndex = getSuccessfulBannerIndex();
+
+                if (successfulBannerIndex !== -1) {
+                    const successfulBanner = banners[successfulBannerIndex];
+                    successfulBanner.show();
+
+                    const trackingObj = {
+                        component: 'banner-picker',
+                        value: successfulBanner.id,
+                    };
+                    ophan.record(trackingObj);
+                }
+
+                if (!results.includes('pending')) {
+                    resolve();
```

Comment:
Do we need to clear the `timeout` at this point? Not sure, just thinking memory leaks

Reply:
The `timeout` has already been cleared at this point. It's being cleared in the `then` block following `banner.canShow`.